### PR TITLE
Add UI and API docs custom context path support

### LINKED ISSUE #12649: Deploying JHipster with a context path breaks SwaggerUI
##### **Overview of the issue**

When deploying JHipster with a context path set, one has to change the Webpack configuration and SwaggerUI according to https://www.jhipster.tech/production/#build, otherwise the client app fails to load.
First off, the doc is incomplete as one also has to prepend the `axios.get` call with the context path, like so: `axios.get("/jhipster/swagger-resources")`. Otherwise the SwaggerUI will not load the Resource definitions.
But even if one follows the doc and the above, the SwaggerUI is still broken.

##### **Motivation for or Use Case**

It should be possible to use a context path when deploying JHipster and use the SwaggerUI without further configuration.

##### **Reproduce the error**

* Generate a new JHipster project
* Set `server.servlet.context-path` to desired context
* Make configuration changes according to the doc and the `axios.get` call
* Run `mvn spring-boot:run` on the project

##### **Related issues**

#5620

##### **Suggest a Fix**

Using relative paths would make it possible to set `server.servlet.context-path` and keep the client app runnable without further configuration. I have already made a commit in my fork for that - see: https://github.com/jhipster/generator-jhipster/compare/main...clocken:running-jhipster-on-a-specified-context

There is still an issue with the SwaggerUI though. If a context path is set, the [Docket](https://github.com/jhipster/jhipster/blob/40fe6c2bdf21de95ce99834652585668f603e347/jhipster-framework/src/main/java/io/github/jhipster/config/apidoc/SpringfoxAutoConfiguration.java#L104) will not generate the correct *OAS*.
To fix that for the `default` group one also has to prepend the context path to `jhipster.api-docs.default-include-pattern`.
But for the `management` group the *OAS* will still not be generated as it matches the api resources by `management.endpoints.web.base-path`.
If I am not mistaken, there is currently no way to specify the correct regex matcher for the Management-[Docket](https://github.com/jhipster/jhipster/blob/40fe6c2bdf21de95ce99834652585668f603e347/jhipster-framework/src/main/java/io/github/jhipster/config/apidoc/SpringfoxAutoConfiguration.java#L146).

This leads me to the following suggestions:
* Use the relative paths in the client app as per my commit above
* First option for fixing the SwaggerUI:
  * Figure out the context path inside the Docket's and prepend them transparently to the `jhipster.api-docs.default-include-pattern` and `management.endpoints.web.base-path`.

This would make it possible to create JHipster deployments with a context path without further configuration.

There's also another solution for the SwaggerUI which I include just for completeness as I don't like it very much.
* Second option for fixing the SwaggerUI:
  * set `jhipster.api-docs.default-include-pattern` to `${spring.servlet.context-path:}/api/.*` and
  * introduce a new variable `jhipster.api-docs.management-include-pattern` to specify the regex for the Management-Docket and set it to `${spring.servlet.context-path:}/management/.*`

I don't like it, because it relies on the correct ordering of the properties so Spring will replace them correctly (an extension to the doc with yet another place to add the context path would do, but defeat the purpose of my request :wink:). Also, this would not work when deploying as a WAR to an Appserver with context path configuration outside of JHipster.

If you are willing to discuss this let me know and I will do the necessary implementations and/or doc updates.

##### **JHipster Version(s)**

Used current main.

```
jhipster-app@0.0.1-SNAPSHOT /home/clocken/jhipster-app
└── generator-jhipster@6.10.4 

```
This maybe a regression - see #5620 and PR #5871. I don't know if the change back to absolute paths was on purpose.

##### **JHipster configuration**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "blueprints": [],
    "otherModules": [],
    "applicationType": "monolith",
    "reactive": false,
    "baseName": "jhipsterApp",
    "jhipsterVersion": "6.10.4",
    "skipClient": false,
    "skipServer": false,
    "skipUserManagement": false,
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "clientPackageManager": "npm",
    "testFrameworks": [],
    "pages": [],
    "creationTimestamp": 1603388507906,
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "en"
    },
    "serviceDiscoveryType": false,
    "authenticationType": "jwt",
    "packageName": "com.mycompany.myapp",
    "serverPort": "8080",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "postgresql",
    "buildTool": "maven",
    "serverSideOptions": [],
    "websocket": false,
    "searchEngine": false,
    "messageBroker": false,
    "enableSwaggerCodegen": false,
    "clientFramework": "angularX",
    "clientTheme": "none",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "packageFolder": "com/mycompany/myapp",
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",
    "clientThemeVariant": "primary",
    "useSass": true,
    "languages": ["en"]
  }
}

</pre>
</details>

##### **Environment and Tools**

```
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment (build 11.0.8+10-post-Ubuntu-0ubuntu118.04.1)
OpenJDK 64-Bit Server VM (build 11.0.8+10-post-Ubuntu-0ubuntu118.04.1, mixed mode, sharing)

git version 2.17.1

node: v12.18.4

npm: 6.14.6

Docker version 19.03.13, build 4484c46d9d
```
##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

ls: no such file or directory: .jhipster/*.json
<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>

##### **Browsers and Operating System**

Unrelated, but I am using Linux and Firefox.

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by PierreBesson**
Depend on #https://github.com/jhipster/jhipster-bom/pull/42
Fixes #12649

These changes follow @clocken's great suggestions and analysis for supporting a custom context-path for JHipster apps including UI and openapi docs support.

### How to test

```
export SERVER_SERVLET_CONTEXT_PATH=/myapp
mvn || gradle
```

TODO:
- [ ] update docs (relating to custom path and war support)

### CODE REVIEW

**Comment by DanielFran** [comment]
@PierreBesson It seems you have conflicts and you need to sign CLA also

**Comment by PierreBesson** [comment]
@clocken Could you sign the CLA please as I have included your commit. Thanks!

**Comment by PierreBesson** [comment]
As for the conflict, I am unable to see it locally :thinking:.

**Comment by pascalgrimaud** [comment]
@PierreBesson : check with your main branch, instead of master branch ;)

**Comment by PierreBesson** [comment]
Holy shit, I forgot that it's not the same.

**Comment by clocken** [comment]
@PierreBesson Thanks for the implementation but it looks like we had a bit of a "race condition" here.
In the meantime the client app has switched to Angular CLI and the `webpack.common.js.ejs` has been removed - see 805e5cca4.
I accounted for that in a copy of your branch and also removed the quotes around the default value for the `management-include-pattern` - see 3ae2d97 and cf2f2aec. Feel free to cherry pick these two commits also.

**Comment by ruddell** [comment]
Before merging, please make sure a path with multiple slashes (such as `/admin/metrics`) is tested both with and without a context-path.  This includes refreshing or directly navigating to the path.  This currently fails if the context-path is relative (like in @clocken's changes).

Without the changes from @clocken, this PR causes the initial page load to fail in prod (attempts to find `/main.js` instead of `/myapp/main.js`).

**Comment by PierreBesson** [comment]
I think the PR is finalized now. Let's validate those cases (including the metrics page as @ruddell pointed out) :

<table>
<tr>
	<th> Framework
	<th> UI without context-path
	<th> UI with context-path
<tr>
	<td> Angular
	<td>
	<td>
<tr>
	<td> React
	<td>
	<td>
<tr>
	<td> Vue
	<td>
	<td>
</table>

Note: the control center should also be checked for context-path support.

**Comment by clocken** [comment]
Ok, I'm done testing now. The `base-href` has to be reverted back to an absolute path and set to a context path manually to be able to page refresh or directly navigate to paths with multiple slashes, changes ready for cherry-picking in [d072af0](https://github.com/clocken/generator-jhipster/commit/d072af0f144f4c08f2f8ac517c3c63ef5a6da36b) @PierreBesson. I also [updated the docs](https://github.com/jhipster/jhipster.github.io/pull/1030) accordingly.
**With that small revert all cases tested positive**:
Framework|UI without context-path|UI with context-path
---------------|------------------------------|--------------------------
Angular|:heavy_check_mark:|:heavy_check_mark:|
React|:heavy_check_mark:|:heavy_check_mark:|
Vue|:heavy_check_mark:|:heavy_check_mark:|

**Comment by PierreBesson** [comment]
I have done extensive testing and rebased to main. I confirm @clocken tests and everything should be fine now. However CI does seem to have issues due to loading jhipster dependencies 7.0.0-beta and not the 7.0.0-SNAPSHOT from the context-path branch. Maybe this is related to the current in-progress release.

@pascalgrimaud if this is not included in the first v7 beta, it can be included in the next beta version.

**Comment by pascalgrimaud** [comment]
yes, I think it's better to include it in beta.1
The beta.0 is for tomorrow, I prefer to secure the release

**Comment by mraible** [comment]
Any updates @PierreBesson?

**Review by mshima — COMMENTED** [review]
index.html moved to https://github.com/jhipster/generator-jhipster/blob/main/generators/client/templates/common/src/main/webapp/index.html.ejs

swagger-ui/index.html moved to https://github.com/jhipster/generator-jhipster/blob/main/generators/client/templates/common/src/main/webapp/swagger-ui/index.html.ejs.

They should be dropped and any changes should be implemented there.

File: generators/generator-constants.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ const GRADLE_VERSION = '6.8.2';
 const JIB_VERSION = '2.7.1';
 
 // Libraries version
-const JHIPSTER_DEPENDENCIES_VERSION = '7.0.0-SNAPSHOT';
+const JHIPSTER_DEPENDENCIES_VERSION = '7.0.0-beta.0';
```
Comment (mshima) [inline generators/generator-constants.js:38]:
```suggestion
const JHIPSTER_DEPENDENCIES_VERSION = '7.0.0-SNAPSHOT';
```

Should be reverted.

**Comment by PierreBesson** [comment]
> index.html moved to https://github.com/jhipster/generator-jhipster/blob/main/generators/client/templates/common/src/main/webapp/index.html.ejs
> 
> swagger-ui/index.html moved to https://github.com/jhipster/generator-jhipster/blob/main/generators/client/templates/common/src/main/webapp/swagger-ui/index.html.ejs.
> 
> They should be dropped and any changes should be implemented there.

Yes I have also noticed this. Thanks !

**Comment by PierreBesson** [comment]
I am currently not able to make this works on my machine. This is not such a big change but it might be impacted by some recent changes in spring boot 2.4, I'm not sure.
I repeat myself but this would be really nice to have for microservices that use a context-path.

**Comment by pascalgrimaud** [comment]
There are several major changes with Spring Boot 2.4:
- all gateway are reactive now
- swagger ui was broken until the fix by @nassimerrahoui, few days ago -> https://github.com/jhipster/generator-jhipster/pull/14029
- then, the swagger ui was still broken because of compression, and has been fixed by @nassimerrahoui few hours ago -> https://github.com/jhipster/generator-jhipster/pull/14204

We had to adapt our code, to change the result from API, it was not easy but I'm confident it's done today.
For adding support for context path, we need to look at these 2 PRs and do similar change. It would be with the same complexity.

Important thing, about context path. You can easily add this property, but there will be several things which are broken:
- npm start, will be broken
- swagger
- logout with OAuth2
- deployment
- etc
 
I forgot probably some items, as I already did this for my customer. And as I said, I'm not sure we can automate this, because the property comes from Spring Boot property and I needed to put this context path in some .vue / .ts file. These files can't guess this property.
So adding this support is difficult IMO :-)

**Comment by DanielFran** [comment]
@PierreBesson @pascalgrimaud Following #14250, this PR should be closed?

**Pull request closed without merging**